thetis/callback.py

```diff
@@ -523,43 +527,120 @@ def __call__(self):
+    """
+    name = 'time integral'
+    variable_names = ['value']
+
+    def __init__(self, spatial_integral, solver_obj, timestepper, name=None, **kwargs):
+        """
+        :arg spatial_integral: Python function that returns an integral in space only
+        :arg solver_obj: Thetis solver object
+        :arg timestepper: Thetis TimeIntegrator object
+        :kwarg name: name of integrated quantity
+        :arg kwargs: any additional keyword arguments, see DiagnosticCallback
+        """
+        if name is not None:
+            self.name = name
+        dt = timestepper.dt
+
+        # Check if implemented then collect quadrature weights and update values at nodes
+        to_do = (
+            timeintegrator.PressureProjectionPicard,
+            timeintegrator.LeapFrogAM3,
+            timeintegrator.SSPRK22ALE,
+        )
+        if isinstance(timestepper, to_do):
+            raise NotImplementedError  # TODO
+
+        elif issubclass(timestepper.__class__, rungekutta.ERKGeneric):
+            weights = timestepper.b
+            updates = timestepper.tendency
+
+        elif issubclass(timestepper.__class__, rungekutta.ERKGenericShuOsher):
+            weights = timestepper.b
+            updates = timestepper.stage_sol
+
+        elif issubclass(timestepper.__class__, rungekutta.DIRKGeneric):
+            weights = timestepper.b
+            updates = timestepper.k
+
+        elif issubclass(timestepper.__class__, rungekutta.DIRKGenericUForm):
+            weights = timestepper.b
+            updates = timestepper.k.copy()
+            updates.insert(0, timestepper.solution_old)
+
+        elif isinstance(timestepper, timeintegrator.ForwardEuler):
+            weights = [1.0, ]
+            updates = [timestepper.solution_old, ]
+
+        elif isinstance(timestepper, timeintegrator.SteadyState):
+            weights = [1.0, ]
+            updates = [timestepper.solution, ]
+
+        elif isinstance(timestepper, timeintegrator.CrankNicolson):
+            theta = solver_obj.options.timestepper_options.implicitness_theta
+            weights = [1 - theta, theta]
+            updates = [timestepper.solution_old, timestepper.solution]
+
+        else:
+            raise ValueError("Timestepper {:s} not recognised.".format(timestepper.__class__.__name__))
+
+        num_updates = len(updates)
+        num_weights = len(weights)
+        try:
+            assert num_updates == num_weights
+        except AssertionError:
+            msg = "Update and weight vectors have different lengths ({:d} vs. {:d})"
+            raise ValueError(msg.format(num_updates, num_weights))
+
+        def time_integral_callback():
+            """Time integrate spatial integral over a single timestep"""
+            integral = 0
+            for i in range(num_weights):
+                stored = i == 0 and hasattr(self, 'previous_update')
+                update_integral = self.previous_update if stored else spatial_integral(updates[i])
```

Comment:
Have you tested this? It should be possible to design a test where the analytical solution is known (constant in time is easy). As far as I can see, `update_integral` is wrong. In some cases `updates` holds the tendencies while, presumably, the `spatial_integral` should be evaluated using the solution itself. What if user wants to evaluate a time integral of a diagnostic field? Then it would not be sufficient to track just the solution. I wonder if it would be better to extend the time integration API to allow for user-defined callback functions do this? The time integrator could then handle time-integration if necessary(?). I would also like to create a better API for user-defined diagnostic fields. `ExportManager.add_export` already takes optional `preproc_func` argument which can be used to compute dianostics but this should be extended. We could think what the API could look like.

Reply:
I am using it for some tsunami simulations at the minute and it works for the integrators I have tried. But no, I haven't done any more extensive testing than that.

Good point - I suppose it would be better to have the time integrators do the quadrature themselves. I will have a think and maybe we can discuss on Friday.